src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,322 @@ namespace System.Net.Sockets
+                if (_currentEngine == null || _currentEngine.IsFull)
                 {
-                    lock (_initLock)
-                    {
-                        if (_engine == null)
-                        {
-                            int port;
-                            Interop.Error err = Interop.Sys.CreateSocketEventPort(&port);
-                            if (err != Interop.Error.SUCCESS)
-                            {
-                                throw new InternalException();
-                            }
+                    _currentEngine = new SocketAsyncEngine();
+                }
 
-                            Interop.Sys.SocketEvent* buffer;
-                            err = Interop.Sys.CreateSocketEventBuffer(EventBufferCount, &buffer);
-                            if (err != Interop.Error.SUCCESS)
-                            {
-                                Interop.Sys.CloseSocketEventPort(port);
-                                throw new InternalException();
-                            }
+                engine = _currentEngine;
+                handle = _currentEngine.AllocateHandle(context);
+            }
+        }
 
-                            var engine = new SocketAsyncEngine(port, buffer);
-                            Task.Factory.StartNew(o =>
-                            {
-                                var eng = (SocketAsyncEngine)o;
-                                for (;;)
-                                {
-                                    try
-                                    {
-                                        eng.EventLoop();
-                                    }
-                                    catch (Exception e)
-                                    {
-                                        Debug.Fail(string.Format("Exception thrown from event loop: {0}", e.Message));
-                                    }
-                                }
-                            }, engine, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
-
-                            Volatile.Write(ref _engine, engine);
-                        }
-                    }
+        private IntPtr AllocateHandle(SocketAsyncContext context)
+        {
+            Debug.Assert(Monitor.IsEntered(_lock));
+            Debug.Assert(!IsFull);
+
+            IntPtr handle = _nextHandle;
+            _handleToContextMap.Add(handle, context);
+
+            _nextHandle = IntPtr.Add(_nextHandle, 1);
+            _outstandingHandles = IntPtr.Add(_outstandingHandles, 1);
+
+            Debug.Assert(handle != ShutdownHandle);
+            return handle;
+        }
+
+        private void FreeHandle(IntPtr handle)
+        {
+            Debug.Assert(handle != ShutdownHandle);
+
+            bool shutdownNeeded = false;
+
+            lock (_lock)
+            {
+                if (_handleToContextMap.Remove(handle))
```

Comment:
Should we assert that Remove returned true?  Or is it ok to free the same handle multiple times?